src/websocket.rs

```diff
@@ -23,64 +24,86 @@ impl AdminWebsocket {
+        Ok(Self {
+            tx,
+            agent_key: None,
+        })
     }
 
     #[instrument(skip(self), err)]
-    pub async fn generate_agent_pubkey(&mut self) -> Result<AgentPubKey> {
+    pub async fn get_agent_key(&mut self) -> Result<AgentPubKey> {
+        if let Some(key) = self.agent_key.clone() {
+            return Ok(key);
+        }
         let response = self.send(AdminRequest::GenerateAgentPubKey).await?;
         match response {
-            AdminResponse::AgentPubKeyGenerated(agent_pubkey) => Ok(agent_pubkey),
+            AdminResponse::AgentPubKeyGenerated(key) => {
+                self.agent_key = Some(key.clone());
+                Ok(key)
+            }
             _ => Err(anyhow!("unexpected response: {:?}", response)),
         }
     }
 
-    // TODO: write and use get_installed_happs
+    #[instrument(skip(self))]
+    pub async fn attach_app_interface(&mut self, happ_port: u16) -> Result<AdminResponse> {
+        info!(port = ?happ_port, "starting app interface");
+        let msg = AdminRequest::AttachAppInterface {
+            port: Some(happ_port),
+        };
+        self.send(msg).await
+    }
+
     #[instrument(skip(self), err)]
-    pub async fn get_installed_happs(&mut self) -> Result<AdminResponse> {
-        todo!()
+    pub async fn list_installed_happs(&mut self) -> Result<Vec<InstalledAppId>> {
+        let response = self.send(AdminRequest::ListActiveApps).await?;
+        match response {
+            AdminResponse::ActiveAppsListed(app_ids) => Ok(app_ids),
+            _ => Err(anyhow!("unexpected response: {:?}", response)),
+        }
     }
 
     #[instrument(
-        skip(self, happ, agent_key, happ_port),
+        skip(self, happ),
         fields(?happ.app_id),
     )]
-    pub async fn install_happ(
-        &mut self,
-        happ: &Happ,
-        agent_key: AgentPubKey,
-        happ_port: u16,
-    ) -> Result<()> {
-        debug!(?agent_key);
-        self.instance_dna_for_agent(happ, agent_key).await?;
-        self.activate_app(&happ.app_id).await?;
-        self.attach_app_interface(happ_port).await?;
+    pub async fn install_happ(&mut self, happ: &Happ) -> Result<()> {
+        if happ.dna_url.is_some() {
+            self.instance_dna_for_agent(happ).await?;
+        } else {
+            debug!(?happ.app_id, "dna_url == None, skipping DNA installation")
+        }
+        self.activate_app(happ).await?;
         info!(?happ.app_id, "installed hApp");
         Ok(())
     }
 
     #[instrument(
         err,
-        skip(self, happ, agent_key),
+        skip(self, happ),
         fields(?happ.app_id)
     )]
-    async fn instance_dna_for_agent(
-        &mut self,
-        happ: &Happ,
-        agent_key: AgentPubKey,
-    ) -> Result<AdminResponse> {
-        let file = crate::download_file(&happ.dna_url)
+    async fn instance_dna_for_agent(&mut self, happ: &Happ) -> Result<AdminResponse> {
+        let agent_key = self
+            .get_agent_key()
```

Comment:
As I see this implementation the idea is to share agent key across all the happs that get installed.  This is correct, but the problem is that it only works for one run of this binary.  The next time it runs a new agent ID will be generated because the key is only stored in RAM.   For the very short term this solution is ok, but it will need to be changed very soon once we decide how we want to manage the Holoport's agent key.  Then we have to check the persisted value, perhaps on the USB stick, and check that it matches an actual key that lives in lair, or something like that.

Reply:
As soon as the keys find their place in HPOS, I'll make this change.